src/mame/brother/lw840.cpp

```diff
@@ -0,0 +1,362 @@
+16 MHz (XT2), 18 MHz (XT1)
+clone of WD37C65
+
+#6
+NKK
+N341256SJ 
+CMOS SRAM (32k x 8)
+
+#8
+Brother/Hitachi
+UC9201-000
+HG71C207FD
+MPU
+208 pins
+H8/3003-alike; can't be H8/3002, because 'Port C' is used in ROM
+H8/300H Advanced Mode
+14.74 MHz CPU Core (XT4), 20 MHz Printer Control (XT3)
+
+#10
+Nippon Steel Semiconductor 
+NN514260J-60
+Fast Page Mode CMOS 256k x 16bit Dynamic RAM
+
+#11
+Macronix
+US3122-A
+64316JW1
+5 Volt 32-Mbit (4M x 8 / 2M x 16) Mask ROM with Page Mode
+
+see https://github.com/BartmanAbyss/brother-hardware/tree/master/4G%20-%20Brother%20LW-840ic for datasheets, manuals, photos
+
+Emulation Status:
+- Keyboard not 100%
+- Screen Refresh Rate not yet measured
+- Printer not working
+
+***************************************************************************/
+
+class gm82c765b_device : public upd765_family_device {
+public:
+	gm82c765b_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+
+	void map(address_map& map) override {
+		map(0x0, 0x0).r(FUNC(gm82c765b_device::msr_r));
+		map(0x1, 0x1).rw(FUNC(gm82c765b_device::fifo_r), FUNC(gm82c765b_device::fifo_w));
+		map(0x2, 0x2).w(FUNC(gm82c765b_device::dor_w));
+		map(0x6, 0x6).rw(FUNC(gm82c765b_device::dma_r), FUNC(gm82c765b_device::dma_w));
+	}
+};
+
+DEFINE_DEVICE_TYPE(GM82C765B, gm82c765b_device, "gm82c765b", "GM82C765B")
+
+gm82c765b_device::gm82c765b_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) : 
+	upd765_family_device(mconfig, GM82C765B, tag, owner, clock)
+{
+	ready_polled = true;
+	ready_connected = false;
+	select_connected = true;
+}
+
+class lw840_state : public driver_device
+{
+public:
+	lw840_state(const machine_config &mconfig, device_type type, const char *tag)
+		: driver_device(mconfig, type, tag),
+		maincpu(*this, "maincpu"),
+		screen(*this, "screen"),
+		fdc(*this, "fdc"),
+		beeper(*this, "beeper"),
+		io_kbrow(*this, "kbrow.%u", 0),
+		rom(*this, "maincpu"),
+		sram(*this, "sram")
+	{ }
+
+	void lw840(machine_config& config);
+
+protected:
+	// driver_device overrides
+	void machine_start() override;
+	void machine_reset() override;
+	void video_start() override;
+
+private:
+	// devices
+	required_device<h83003_device> maincpu;
+	required_device<screen_device> screen;
+	required_device<gm82c765b_device> fdc;
+	required_device<beep_device> beeper;
+	required_ioport_array<9> io_kbrow;
+	required_region_ptr<uint16_t> rom;
+	required_shared_ptr<uint16_t> sram;
+
+	uint8_t keyboard{};
+	uint8_t irq_toggle{};
+
+	// screen updates
+	uint32_t screen_update(screen_device& screen, bitmap_rgb32& bitmap, const rectangle& cliprect);
+
+	uint16_t port7_r(offs_t, uint16_t mem_mask = ~0) {
+		auto row = keyboard & 0x0f;
+		if(row <= 8)
+			return io_kbrow[row]->read();
+
+		// seems to be able to control power-on self test if not 0xff
+		return 0xffff & mem_mask;
+	}
+	uint16_t keyboard_r(offs_t, uint16_t mem_mask = ~0) {
+		return (keyboard << 8) & mem_mask; 
+	}
```

Comment:
As I mentioned before, the `mem_mask` is for lane select.  You don’t need to check it in cases like this where the result doesn’t depend on which lanes are selected.  You only need to check it if different peripherals are selected depending on the selected lanes or something,